cmd/juju/commands/switch.go

```diff
@@ -101,25 +101,94 @@ func (c *switchCommand) SetFlags(f *gnuflag.FlagSet) {
 }
 
 func (c *switchCommand) Init(args []string) error {
-	var err error
-	var target string
 	if c.modelName != "" && c.controllerName != "" {
 		return errors.Trace(fmt.Errorf("cannot specify both a --model and --controller"))
 	}
-	if c.modelName != "" || c.controllerName != "" {
-		// translate `-c ctrl` to `ctrl:` and `-m model` to `:model`
-		target = fmt.Sprintf("%s:%s", c.controllerName, c.modelName)
-		err = cmd.CheckEmpty(args)
+
+	if c.controllerName != "" {
+		err := cmd.CheckEmpty(args)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	} else if c.modelName != "" {
+		err := cmd.CheckEmpty(args)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// This means we can support arguemets like `juju switch -m mycontroller:mymodel`
+		c.parseModelName(c.modelName)
 	} else {
-		target, err = cmd.ZeroOrOneArgs(args)
+		target, err := cmd.ZeroOrOneArgs(args)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// if the target does not contain a ":", it is an ambiguous target so
+		// we cannot parse it as a model name
+		if strings.Contains(target, ":") {
+			c.parseModelName(target)
+		} else {
+			c.controllerOrModelName = target
+		}
 	}
-	if err != nil {
-		return err
+
+	// expand "-" syntactic sugar where it's found
+	if c.controllerName == "-" {
+		previous, _, err := c.Store.PreviousController()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.controllerName = previous
 	}
-	// handling special arguments like -:, -, and :- is done as part of initialization to outline their
-	// "syntaxic sugar" kind
-	c.Target, err = c.resolvePreviousTarget(target)
-	return err
+
+	if c.modelName == "-" {
+		controller, err := c.Store.CurrentController()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		previous, err := c.Store.PreviousModel(controller)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.modelName = previous
+	}
+
+	if c.controllerOrModelName == "-" {
+		c.controllerOrModelName = ""
+		currentController, err := c.Store.CurrentController()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		previousController, changed, err := c.Store.PreviousController()
+		if errors.Is(err, errors.NotFound) || !changed {
+			previousController = currentController
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+		c.controllerName = previousController
+
+		// if the last switch was intra-controller, we need to figure out from which model
+		// and switch back there. Otherwise (i.e. inter-controller), it is sufficient to
+		// switch to just the previous controller.
+		if currentController == previousController {
+			previousModel, err := c.Store.PreviousModel(currentController)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			c.modelName = previousModel
+		}
```

Comment:
**suggestion:** We can simplify a bit more:
```go
                c.controllerOrModelName = ""
		previousController, changed, err := c.Store.PreviousController()
		if err != nil && !errors.Is(err, errors.NotFound) {
		     return errors.Trace(err)
		}

		// if the last switch was intra-controller, we need to figure out from which model
		// and switch back there. Otherwise (i.e. inter-controller), it is sufficient to
		// switch to just the previous controller.		
		if changed {
                       c.controllerName = previousController		     
		} else {
		        // FIXME : this code shouldn't need to be necessary, since empty controllerName will be setted to current... Unless we choose to do a full resolution in init and NEVER have a model without a controller.
		        // c.controllerName, err := c.Store.CurrentController()
		        // if err != nil {
			//         return errors.Trace(err)
		        // }
			previousModel, err := c.Store.PreviousModel(currentController)
			if err != nil {
				return errors.Trace(err)
			}
			c.modelName = previousModel
		}
```__

Reply:
We still need to get the current controller to run `PreviousModel`

So we may as well set it in my opinion. The more explicit the better